lua/coq/init.lua

```diff
@@ -0,0 +1,136 @@
+    if not (code == 0 or code == 143) then
+        err_exit = true
+        api.nvim_err_writeln(msg)
+    else
+        err_exit = false
+    end
+    job_id = nil
+end
+
+local on_stdout =
+  function(_, msg) api.nvim_out_write(table.concat(msg, linesep)) end
+
+local on_stderr =
+  function(_, msg) api.nvim_err_write(table.concat(msg, linesep)) end
+
+local go, _py3 = pcall(api.nvim_get_var, 'python3_host_prog')
+local py3 = go and _py3 or (is_win and 'python' or 'python3')
+
+local main = function()
+    local v_py = cwd .. (is_win and [[/.vars/runtime/Scripts/python.exe]] or
+                   '/.vars/runtime/bin/python3')
+    local win_proxy = cwd .. [[/win.bat]]
+
+    if is_win then
+        if fn.filereadable(v_py) == 1 then
+            return {v_py}
+        else
+            return {win_proxy, py3}
+        end
+    else
+        if fn.filereadable(v_py) == 1 then
+            return {v_py}
+        else
+            return {py3}
+        end
+    end
+end
+
+local start = function(deps, ...)
+    local args = v.tbl_flatten {deps and py3 or main(), {'-m', 'coq'}, {...}}
+
+    local params = {
+        cwd = cwd,
+        on_exit = on_exit,
+        on_stdout = on_stdout,
+        on_stderr = on_stderr,
+    }
+    job_id = fn.jobstart(args, params)
+    return job_id
+end
+
+M.COQdeps = function() start(true, 'deps') end
+
+local set_coq_call = function(cmd)
+    M[cmd] = function(...)
+        local args = {...}
+
+        if not job_id then
+            local server = api.nvim_call_function('serverstart', {})
+            job_id = start(false, 'run', '--socket', server)
+        end
+
+        if not err_exit and _G[cmd] then
+            _G[cmd](args)
+        else
+            v.defer_fn(function()
+                if err_exit then
+                    return
+                else
+                    M[cmd](unpack(args))
+                end
+            end, POLLING_RATE)
+        end
+    end
+end
+
+api.nvim_command [[command! -nargs=0 COQdeps lua require("coq").COQdeps()]]
+
+set_coq_call('COQnow')
+api.nvim_command [[command! -complete=customlist,coq#complete_now -nargs=* COQnow lua require("coq").COQnow(<f-args>)]]
+
+set_coq_call('COQstats')
+api.nvim_command [[command! -nargs=* COQstats lua require("coq").COQstats(<f-args>)]]
+
+set_coq_call('COQhelp')
+api.nvim_command [[command! -complete=customlist,coq#complete_help  -nargs=* COQhelp lua require("coq").COQhelp(<f-args>)]]
+
+M.lsp_ensure_capabilities = function(cfg)
+    local spec1 = {capabilities = v.lsp.protocol.make_client_capabilities()}
+    local spec2 = {
+        capabilities = {
+            textDocument = {
+                completion = {completionItem = {snippetSupport = true}},
+            },
+        },
+    }
+    local maps = (cfg or {}).capabilities and {spec2} or {spec1, spec2}
+    local new = v.tbl_deep_extend('force', cfg or v.empty_dict(), unpack(maps))
+    return new
+end
+
+M.setup = function(opts)
```

Comment:
same thing about backwards compatibility here.

as long as I keep `coq_settings` around, this function unfortunately will redundant.

I don't see the beenfits of breaking many users existing configuration worth it, since most old style vim plugins use global variables for ease of compatibility with vimscript.

Reply:
i am really sorry I have to comment these things, I want you to know that I really do appreciate your input, and I hate deleting other peoples code.

Reply:
and dont worry about any changes I ask here. I am just commenting so you know why I overwrote your code, not that I am asking you to make the changes.

I will do that myself, I just hate being the person who has to tell other people no. If I have to deny something, I want to make sure the other person 100% understand where I am coming from

Reply:
Purpose it to make sure it lazy loads when the user wants it to. Previous solution will start coq just because you import the module. So all the variables has to be set before require('coq'). When someone fire up neovim doesn't mean he wants to use the completion. They may be using two more than one LSP plugins for example coc and builtin lsp together. So this only needs to be ran when the built-in lsp are setup.

Reply:
Although, user can workaround to set the configuration first. It may make more sense to avoid this considering the backward compatibility

Reply:
Actually the previous code was already lazy loading everything, with the exception of `auto_start`, which was impossible to avoid.

Thats why you had to call `COQnow`.

anyways though, this PR cannot go in until I test for windows compat

Reply:
That's why git introduced branches. 🙂 Create a feature branch. Do your backward compatibility stuff. Let people do the testing. One person can only do so much.

Reply:
yub, im making a windows vm :)